controllers/storageconsumer/storageconsumer_upgrade_controller.go

```diff
@@ -0,0 +1,218 @@
+)
+
+// StorageConsumerReconciler reconciles a StorageConsumer object
+type StorageConsumerUpgradeReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *StorageConsumerUpgradeReconciler) SetupWithManager(mgr ctrl.Manager) error {
+
+	operatorVersionPredicate := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			obj := e.Object.(*ocsv1alpha1.StorageConsumer)
+			return obj != nil && strings.HasPrefix(obj.Status.Client.OperatorVersion, "4.18")
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return false
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return false
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			return false
+		},
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&ocsv1alpha1.StorageConsumer{}, builder.WithPredicates(operatorVersionPredicate)).
+		Complete(r)
+}
+
+// +kubebuilder:rbac:groups=ocs.openshift.io,resources=storageconsumers,verbs=get;watch;create;update
+// +kubebuilder:rbac:groups=core,resources=configmaps,verbs=get;create;update
+
+func (r *StorageConsumerUpgradeReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+
+	log := ctrl.LoggerFrom(ctx)
+
+	storageConsumer := &ocsv1alpha1.StorageConsumer{}
+	storageConsumer.Name = request.Name
+	storageConsumer.Namespace = request.Namespace
+
+	if err := r.Client.Get(ctx, client.ObjectKeyFromObject(storageConsumer), storageConsumer); errors.IsNotFound(err) {
+		log.Info("No StorageConsumer resource.")
+		// Request object not found, could have been deleted after reconcile request.
+		// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+		// Return and don't requeue
+		return reconcile.Result{}, nil
+	} else if err != nil {
+		// Error reading the object - requeue the request.
+		log.Error(err, "Failed to retrieve StorageConsumer.")
+		return reconcile.Result{}, err
+	}
+
+	if !storageConsumer.GetDeletionTimestamp().IsZero() {
+		return reconcile.Result{}, nil
+	}
+
+	if storageConsumer.Spec.ResourceNameMappingConfigMap.Name != "" {
+		return reconcile.Result{}, nil
+	}
+
+	storageCluster, err := util.GetStorageClusterInNamespace(ctx, r.Client, storageConsumer.Namespace)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	availableServices, err := util.GetAvailableServices(ctx, r.Client, storageCluster)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to get available services configured in StorageCluster: %v", err)
+	}
+
+	consumerConfigMap := &corev1.ConfigMap{}
+	consumerConfigMap.Name = fmt.Sprintf("storageconsumer-%v", util.FnvHash(storageConsumer.Name))
+	consumerConfigMap.Namespace = storageConsumer.Namespace
+
+	data := util.GetStorageConsumerDefaultResourceNames(
+		storageConsumer.Name,
+		string(storageConsumer.UID),
+		availableServices,
+	)
+	resourceMap := util.WrapStorageConsumerResourceMap(data)
```

Comment:
Please move this wrapper creation directly above the first line it is being used.